x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java

```diff
@@ -0,0 +1,648 @@
+            onFileChanged(file);
+        }
+
+        @Override
+        public void onFileChanged(Path file) {
+            try {
+                onChange.run();
+            } catch (Exception e) {
+                logger.warn(new ParameterizedMessage("An error occurred while reloading file {}", file), e);
+            }
+        }
+    }
+
+    /**
+     * Creates a new {@link PrivilegedResourceRetriever} to be used with the {@link IDTokenValidator} by passing the
+     * necessary client SSLContext and hostname verification configuration
+     */
+    private PrivilegedResourceRetriever getPrivilegedResourceRetriever() {
+        final String sslKey = RealmSettings.realmSslPrefix(realmConfig.identifier());
+        final SSLConfiguration sslConfiguration = sslService.getSSLConfiguration(sslKey);
+        boolean isHostnameVerificationEnabled = sslConfiguration.verificationMode().isHostnameVerificationEnabled();
+        final HostnameVerifier verifier = isHostnameVerificationEnabled ? new DefaultHostnameVerifier() : NoopHostnameVerifier.INSTANCE;
+        return new PrivilegedResourceRetriever(sslService.sslContext(sslConfiguration), verifier, realmConfig);
+    }
+
+    static class PrivilegedResourceRetriever extends DefaultResourceRetriever {
+        private SSLContext clientContext;
+        private HostnameVerifier verifier;
+        private RealmConfig config;
+
+        PrivilegedResourceRetriever(final SSLContext clientContext, final HostnameVerifier verifier, final RealmConfig config) {
+            super();
+            this.clientContext = clientContext;
+            this.verifier = verifier;
+            this.config = config;
+        }
+
+        @Override
+        public Resource retrieveResource(final URL url) throws IOException {
+            SpecialPermission.check();
+            try {
+                return AccessController.doPrivileged(
+                    (PrivilegedExceptionAction<Resource>) () -> {
+                        final BasicHttpContext context = new BasicHttpContext();
+                        final RequestConfig requestConfig = RequestConfig.custom()
+                            .setConnectTimeout(Math.toIntExact(config.getSetting(HTTP_CONNECT_TIMEOUT).getMillis()))
+                            .setConnectionRequestTimeout(Math.toIntExact(config.getSetting(HTTP_CONNECTION_READ_TIMEOUT).getSeconds()))
+                            .setSocketTimeout(Math.toIntExact(config.getSetting(HTTP_SOCKET_TIMEOUT).getMillis())).build();
+                        try (CloseableHttpClient client = HttpClients.custom()
+                            .setSSLContext(clientContext)
+                            .setSSLHostnameVerifier(verifier)
+                            .setDefaultRequestConfig(requestConfig)
+                            .build()) {
+                            HttpGet get = new HttpGet(url.toURI());
+                            HttpResponse response = client.execute(get, context);
```

Comment:
I haven't tested this at all but something like this should do the trick:

```diff
diff --git a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java
index d7af96ba08b..a3503ac2e92 100644
--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java
@@ -7,11 +7,14 @@ package org.elasticsearch.xpack.security.authc.oidc;
 
 import com.nimbusds.jose.JOSEException;
 import com.nimbusds.jose.JWSAlgorithm;
+import com.nimbusds.jose.jwk.JWK;
+import com.nimbusds.jose.jwk.JWKSelector;
 import com.nimbusds.jose.jwk.JWKSet;
+import com.nimbusds.jose.jwk.source.JWKSource;
 import com.nimbusds.jose.proc.BadJOSEException;
-import com.nimbusds.jose.util.DefaultResourceRetriever;
+import com.nimbusds.jose.proc.JWSVerificationKeySelector;
+import com.nimbusds.jose.proc.SecurityContext;
 import com.nimbusds.jose.util.IOUtils;
-import com.nimbusds.jose.util.Resource;
 import com.nimbusds.jwt.JWT;
 import com.nimbusds.jwt.JWTClaimsSet;
 import com.nimbusds.oauth2.sdk.AuthorizationCode;
@@ -51,15 +54,12 @@ import org.apache.http.conn.ssl.DefaultHostnameVerifier;
 import org.apache.http.conn.ssl.NoopHostnameVerifier;
 import org.apache.http.entity.ContentType;
 import org.apache.http.impl.auth.BasicScheme;
-import org.apache.http.impl.client.CloseableHttpClient;
-import org.apache.http.impl.client.HttpClients;
 import org.apache.http.impl.nio.client.CloseableHttpAsyncClient;
 import org.apache.http.impl.nio.client.HttpAsyncClients;
 import org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager;
 import org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor;
 import org.apache.http.message.BasicNameValuePair;
 import org.apache.http.nio.reactor.ConnectingIOReactor;
-import org.apache.http.protocol.BasicHttpContext;
 import org.apache.http.util.EntityUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
@@ -72,6 +72,8 @@ import org.elasticsearch.common.Nullable;
 import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.SuppressForbidden;
 import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.util.concurrent.EsExecutors;
+import org.elasticsearch.common.util.concurrent.ListenableFuture;
 import org.elasticsearch.watcher.FileChangesListener;
 import org.elasticsearch.watcher.FileWatcher;
 import org.elasticsearch.watcher.ResourceWatcherService;
@@ -85,6 +87,7 @@ import javax.net.ssl.SSLContext;
 import java.io.IOException;
 import java.io.UnsupportedEncodingException;
 import java.net.URI;
+import java.net.URISyntaxException;
 import java.net.URL;
 import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
@@ -97,10 +100,11 @@ import java.text.ParseException;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
 
 import static org.elasticsearch.xpack.core.security.authc.oidc.OpenIdConnectRealmSettings.ALLOWED_CLOCK_SKEW;
-import static org.elasticsearch.xpack.core.security.authc.oidc.OpenIdConnectRealmSettings.HTTP_CONNECT_TIMEOUT;
 import static org.elasticsearch.xpack.core.security.authc.oidc.OpenIdConnectRealmSettings.HTTP_CONNECTION_READ_TIMEOUT;
+import static org.elasticsearch.xpack.core.security.authc.oidc.OpenIdConnectRealmSettings.HTTP_CONNECT_TIMEOUT;
 import static org.elasticsearch.xpack.core.security.authc.oidc.OpenIdConnectRealmSettings.HTTP_MAX_CONNECTIONS;
 import static org.elasticsearch.xpack.core.security.authc.oidc.OpenIdConnectRealmSettings.HTTP_MAX_ENDPOINT_CONNECTIONS;
 import static org.elasticsearch.xpack.core.security.authc.oidc.OpenIdConnectRealmSettings.HTTP_SOCKET_TIMEOUT;
@@ -129,7 +133,7 @@ public class OpenIdConnectAuthenticator {
         this.rpConfig = rpConfig;
         this.sslService = sslService;
         this.httpClient = createHttpClient();
-        this.idTokenValidator = createIdTokenValidator(getPrivilegedResourceRetriever());
+        this.idTokenValidator = createIdTokenValidator();
         this.watcherService = watcherService;
     }
 
@@ -217,7 +221,18 @@ public class OpenIdConnectAuthenticator {
             } else {
                 claimsListener.onResponse(verifiedIdTokenClaims);
             }
-        } catch (com.nimbusds.oauth2.sdk.ParseException | JOSEException | BadJOSEException e) {
+        } catch (BadJOSEException e) {
+            if ("Signed JWT rejected: Another algorithm expected, or no matching key(s) found".equals(e.getMessage()) &&
+                opConfig.getJwkSetPath().startsWith("https://")) {
+                ((ReloadableJWKSource) ((JWSVerificationKeySelector) idTokenValidator.getJWSKeySelector()).getJWKSource())
+                    .triggerReload(ActionListener.wrap(v -> {
+                        // redo
+                        getUserClaims(accessToken, idToken, expectedNonce, claimsListener); // FIXME need to avoid continuous loop here
+                    }, claimsListener::onFailure));
+            } else {
+                claimsListener.onFailure(new ElasticsearchSecurityException("Failed to parse or validate the ID Token", e));
+            }
+        } catch (com.nimbusds.oauth2.sdk.ParseException | JOSEException e) {
             claimsListener.onFailure(new ElasticsearchSecurityException("Failed to parse or validate the ID Token", e));
         }
     }
@@ -521,7 +536,7 @@ public class OpenIdConnectAuthenticator {
     /*
      * Creates an {@link IDTokenValidator} based on the current Relying Party configuration
      */
-    IDTokenValidator createIdTokenValidator(final PrivilegedResourceRetriever retriever) {
+    IDTokenValidator createIdTokenValidator() {
         try {
             final JWSAlgorithm requestedAlgorithm = rpConfig.getSignatureAlgorithm();
             final int allowedClockSkew = Math.toIntExact(realmConfig.getSetting(ALLOWED_CLOCK_SKEW).getMillis());
@@ -534,10 +549,9 @@ public class OpenIdConnectAuthenticator {
                 String jwkSetPath = opConfig.getJwkSetPath();
                 if (jwkSetPath.startsWith("https://")) {
                     idTokenValidator = new IDTokenValidator(opConfig.getIssuer(), rpConfig.getClientId(),
-                        requestedAlgorithm, new URL(jwkSetPath), retriever);
-
+                        new JWSVerificationKeySelector(requestedAlgorithm, new ReloadableJWKSource(new URL(jwkSetPath))), null);
                 } else {
-                    setMetadataFileWatcher(jwkSetPath, retriever);
+                    setMetadataFileWatcher(jwkSetPath);
                     final JWKSet jwkSet = readJwkSetFromFile(jwkSetPath);
                     idTokenValidator = new IDTokenValidator(opConfig.getIssuer(), rpConfig.getClientId(), requestedAlgorithm, jwkSet);
                 }
@@ -549,10 +563,10 @@ public class OpenIdConnectAuthenticator {
         }
     }
 
-    private void setMetadataFileWatcher(String jwkSetPath, PrivilegedResourceRetriever resourceRetriever) throws IOException {
+    private void setMetadataFileWatcher(String jwkSetPath) throws IOException {
         final Path path = realmConfig.env().configFile().resolve(jwkSetPath);
         FileWatcher watcher = new FileWatcher(path);
-        watcher.addListener(new FileListener(logger, () -> this.idTokenValidator = createIdTokenValidator(resourceRetriever)));
+        watcher.addListener(new FileListener(logger, () -> this.idTokenValidator = createIdTokenValidator()));
         watcherService.add(watcher, ResourceWatcherService.Frequency.MEDIUM);
     }
 
@@ -594,54 +608,70 @@ public class OpenIdConnectAuthenticator {
         }
     }
 
-    /**
-     * Creates a new {@link PrivilegedResourceRetriever} to be used with the {@link IDTokenValidator} by passing the
-     * necessary client SSLContext and hostname verification configuration
-     */
-    private PrivilegedResourceRetriever getPrivilegedResourceRetriever() {
-        final String sslKey = RealmSettings.realmSslPrefix(realmConfig.identifier());
-        final SSLConfiguration sslConfiguration = sslService.getSSLConfiguration(sslKey);
-        boolean isHostnameVerificationEnabled = sslConfiguration.verificationMode().isHostnameVerificationEnabled();
-        final HostnameVerifier verifier = isHostnameVerificationEnabled ? new DefaultHostnameVerifier() : NoopHostnameVerifier.INSTANCE;
-        return new PrivilegedResourceRetriever(sslService.sslContext(sslConfiguration), verifier, realmConfig);
-    }
+    private class ReloadableJWKSource<C extends SecurityContext> implements JWKSource<C> {
+
+        private final AtomicReference<ListenableFuture<Void>> reloadFutureRef = new AtomicReference<>();
+        private final URL jwkSetPath;
 
-    static class PrivilegedResourceRetriever extends DefaultResourceRetriever {
-        private SSLContext clientContext;
-        private HostnameVerifier verifier;
-        private RealmConfig config;
+        private volatile JWKSet retrievedJWKSet = new JWKSet();
 
-        PrivilegedResourceRetriever(final SSLContext clientContext, final HostnameVerifier verifier, final RealmConfig config) {
-            super();
-            this.clientContext = clientContext;
-            this.verifier = verifier;
-            this.config = config;
+        private ReloadableJWKSource(URL jwkSetPath) {
+            this.jwkSetPath = jwkSetPath;
         }
 
         @Override
-        public Resource retrieveResource(final URL url) throws IOException {
-            SpecialPermission.check();
+        public List<JWK> get(JWKSelector jwkSelector, C context) {
+            return jwkSelector.select(retrievedJWKSet);
+        }
+
+        void triggerReload(ActionListener<Void> toNotify) {
+            ListenableFuture<Void> future = reloadFutureRef.get();
+            if (future == null) {
+                while (future == null) {
+                    future = new ListenableFuture<>();
+                    if (reloadFutureRef.compareAndSet(null, future)) {
+                        reloadAsync(future);
+                    } else {
+                        future = reloadFutureRef.get();
+                    }
+                }
+            }
+            future.addListener(toNotify, EsExecutors.newDirectExecutorService(), null);
+        }
+
+        void reloadAsync(final ListenableFuture<Void> future) {
             try {
-                return AccessController.doPrivileged(
-                    (PrivilegedExceptionAction<Resource>) () -> {
-                        final BasicHttpContext context = new BasicHttpContext();
-                        final RequestConfig requestConfig = RequestConfig.custom()
-                            .setConnectTimeout(Math.toIntExact(config.getSetting(HTTP_CONNECT_TIMEOUT).getMillis()))
-                            .setConnectionRequestTimeout(Math.toIntExact(config.getSetting(HTTP_CONNECTION_READ_TIMEOUT).getSeconds()))
-                            .setSocketTimeout(Math.toIntExact(config.getSetting(HTTP_SOCKET_TIMEOUT).getMillis())).build();
-                        try (CloseableHttpClient client = HttpClients.custom()
-                            .setSSLContext(clientContext)
-                            .setSSLHostnameVerifier(verifier)
-                            .setDefaultRequestConfig(requestConfig)
-                            .build()) {
-                            HttpGet get = new HttpGet(url.toURI());
-                            HttpResponse response = client.execute(get, context);
-                            return new Resource(IOUtils.readInputStreamToString(response.getEntity().getContent(),
-                                StandardCharsets.UTF_8), response.getEntity().getContentType().getValue());
+                final HttpGet httpGet = new HttpGet(jwkSetPath.toURI());
+                AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+                    httpClient.execute(httpGet, new FutureCallback<HttpResponse>() {
+                        @Override
+                        public void completed(HttpResponse result) {
+                            try {
+                                retrievedJWKSet =
+                                    JWKSet.parse(IOUtils.readInputStreamToString(result.getEntity().getContent(), StandardCharsets.UTF_8));
+                                reloadFutureRef.set(null);
+                            } catch (IOException | ParseException e) {
+                                failed(e);
+                            }
+                        }
+
+                        @Override
+                        public void failed(Exception ex) {
+                            future.onFailure(new ElasticsearchSecurityException("Failed to get JWK Set.", ex));
+                            reloadFutureRef.set(null);
+                        }
+
+                        @Override
+                        public void cancelled() {
+                            future.onFailure(new ElasticsearchSecurityException("Failed to get JWK Set. Request was cancelled"));
+                            reloadFutureRef.set(null);
                         }
                     });
-            } catch (final PrivilegedActionException e) {
-                throw (IOException) e.getCause();
+                    return null;
+                });
+            } catch (URISyntaxException e) {
+                future.onFailure(e);
+                reloadFutureRef.set(null);
             }
         }
     }
```

Reply:
Awesome, thanks for that ! I spent some time trying to add the retry logic in `public List<JWK> get(JWKSelector jwkSelector, C context)` instead of the validation method so that we also get the `kid` and check if it is contained in the currently cached JWKSet ( similar to what they do in their `RemoteJWKSource` - but I stumbled and fell and decided it's not that important of an enhanecement ( if any ).